File: src/lib/git_manager.py
"""Git repository manager for Zephyr Desktop.

Provides clone, validation, info retrieval, and recent commit listing
using gitpython. Used by the loop runner to validate repos before
starting containers.
"""

import logging
from pathlib import Path
from typing import Callable

import git
from git.exc import InvalidGitRepositoryError, NoSuchPathError

logger = logging.getLogger("zephyr.git")


class GitManager:
    """Manages Git repository operations for Zephyr projects.

    Wraps gitpython to provide clone, validation, info, and commit
    history functionality needed by the loop execution engine.
    """

    def clone_repo(
        self,
        url: str,
        target_dir: Path,
        progress_callback: Callable | None = None,
    ) -> Path:
        """Clone a remote repository to a local directory.

        Args:
            url: Remote repository URL (HTTPS or SSH).
            target_dir: Local path to clone into. Must not already exist
                as a non-empty directory.
            progress_callback: Optional callable receiving progress strings
                from the clone operation.

        Returns:
            Path to the cloned repository root.

        Raises:
            ValueError: If *url* is empty.
            FileExistsError: If *target_dir* already exists and is non-empty.
            git.exc.GitCommandError: On clone failure (network, auth, etc.).
        """
        if not url or not url.strip():
            raise ValueError("Repository URL must not be empty")

        target = Path(target_dir)
        if target.exists() and any(target.iterdir()):
            raise FileExistsError(
                f"Target directory already exists and is non-empty: {target}"
            )

        logger.info("Cloning %s -> %s", url, target)

        progress = _ProgressHandler(progress_callback) if progress_callback else None

        repo = git.Repo.clone_from(
            url,
            str(target),
            progress=progress,
        )

        logger.info("Clone complete: %s", repo.working_dir)
        return Path(repo.working_dir)

    def validate_repo(self, path: Path) -> bool:
        """Check whether *path* is a valid Git repository.

        Args:
            path: Directory to check.

        Returns:
            True if *path* is a valid Git repo, False otherwise.
        """
        try:
            repo = git.Repo(str(path))
            # Access git_dir to confirm it's truly valid
            _ = repo.git_dir
            return True
        except (InvalidGitRepositoryError, NoSuchPathError):
            return False
        except Exception:
            return False

    def get_repo_info(self, path: Path) -> dict:
        """Return metadata about the repository at *path*.

        Args:
            path: Root of a Git repository.

        Returns:
            Dict with keys: ``branch`` (current branch name or detached
            HEAD hash), ``last_commit`` (short hash of HEAD), ``remote_url``
            (URL of 'origin' remote, or empty string).

        Raises:
            InvalidGitRepositoryError: If *path* is not a Git repo.
        """
        repo = git.Repo(str(path))

        # Current branch
        try:
            branch = repo.active_branch.name
        except TypeError:
            # Detached HEAD — fall back to commit hash
            try:
                branch = str(repo.head.commit)[:12]
            except ValueError:
                branch = ""

        # Last commit hash
        try:
            last_commit = str(repo.head.commit)[:12]
        except ValueError:
            # Empty repo with no commits
            last_commit = ""

        # Origin remote URL
        remote_url = ""
        try:
            if "origin" in [r.name for r in repo.remotes]:
                remote_url = repo.remotes.origin.url
        except Exception:
            pass

        return {
            "branch": branch,
            "last_commit": last_commit,
            "remote_url": remote_url,
        }

    def get_recent_commits(self, path: Path, count: int = 10) -> list[dict]:
        """Return the *count* most recent commits from HEAD.

        Args:
            path: Root of a Git repository.
            count: Maximum number of commits to return.

        Returns:
            List of dicts, each with keys: ``hash`` (full SHA), ``message``
            (first line of commit message), ``author`` (name), ``date``
            (ISO 8601 string).

        Raises:
            InvalidGitRepositoryError: If *path* is not a Git repo.
        """
        repo = git.Repo(str(path))

        commits: list[dict] = []
        try:
            for commit in repo.iter_commits(max_count=count):
                commits.append(
                    {
                        "hash": str(commit.hexsha),
                        "message": commit.message.strip().split("\n")[0],
                        "author": str(commit.author),
                        "date": commit.committed_datetime.isoformat(),
                    }
                )
        except ValueError:
            # Empty repo — no commits to iterate
            pass

        return commits


class _ProgressHandler(git.RemoteProgress):
    """Translates gitpython clone progress into callback invocations."""

    def __init__(self, callback: Callable) -> None:
        super().__init__()
        self._callback = callback

    def update(
        self, op_code: int, cur_count, max_count=None, message: str = ""
    ) -> None:
        """Called by gitpython during remote operations."""
        if message:
            self._callback(message)
        elif max_count:
            self._callback(f"{cur_count}/{max_count}")

File: src/services/git-manager.ts
/**
 * Git repository manager for Zephyr Desktop.
 *
 * Provides clone, validation, info retrieval, and recent commit listing
 * using simple-git. Used by the loop runner to validate repos before
 * starting containers.
 */

import { SimpleGit, simpleGit } from 'simple-git';
import path from 'path';
import { existsSync, readdirSync, mkdirSync, rmSync } from 'fs';

/**
 * Type definitions for Git operations
 */
export interface RepoInfo {
  /** Current branch name or detached HEAD hash */
  branch: string;
  /** URL of the 'origin' remote, empty string if not found */
  remoteUrl: string;
  /** Whether the working directory has uncommitted changes */
  isDirty: boolean;
}

export interface Commit {
  /** Full commit SHA hash */
  hash: string;
  /** First line of commit message */
  message: string;
  /** Author name and email */
  author: string;
  /** ISO 8601 timestamp */
  date: string;
}

/**
 * Progress callback for clone operations.
 * Called periodically during git clone with progress information.
 */
export type CloneProgressCallback = (progress: string) => void;

/**
 * Manages Git repository operations for Zephyr projects.
 *
 * Wraps simple-git to provide clone, validation, info, and commit
 * history functionality needed by the loop execution engine.
 */
export class GitManager {
  private git: SimpleGit;

  constructor() {
    this.git = simpleGit();
  }

  /**
   * Clone a remote repository to a local directory.
   *
   * @param url - Remote repository URL (HTTPS or SSH)
   * @param dest - Local path to clone into. Must not already exist as a non-empty directory
   * @param onProgress - Optional callback receiving progress strings from the clone operation
   * @throws Error if clone fails (network, auth, invalid URL, etc.)
   * @throws Error if destination directory already exists and is non-empty
   */
  async cloneRepo(
    url: string,
    dest: string,
    onProgress?: CloneProgressCallback
  ): Promise<void> {
    if (!url || !url.trim()) {
      throw new Error('Repository URL must not be empty');
    }

    const destPath = path.resolve(dest);

    // Check if destination exists and is non-empty
    if (existsSync(destPath)) {
      const files = readdirSync(destPath);
      if (files.length > 0) {
        throw new Error(
          `Target directory already exists and is non-empty: ${destPath}`
        );
      }
    }

    // Ensure parent directory exists
    const parentDir = path.dirname(destPath);
    if (!existsSync(parentDir)) {
      mkdirSync(parentDir, { recursive: true });
    }

    try {
      // Clone the repository
      // Note: simple-git progress callbacks work differently than gitpython
      // We receive a ProgressSummary object during clone
      await this.git.clone(url, destPath);

      if (onProgress) {
        onProgress('Clone complete');
      }
    } catch (error) {
      // Clean up the destination directory if clone failed
      if (existsSync(destPath)) {
        try {
          rmSync(destPath, { recursive: true, force: true });
        } catch {
          // Ignore cleanup errors
        }
      }
      throw error;
    }
  }

  /**
   * Check whether a path is a valid Git repository.
   *
   * @param repoPath - Directory to check
   * @returns True if the path is a valid Git repo, false otherwise
   */
  async validateRepo(repoPath: string): Promise<boolean> {
    try {
      const gitDir = path.resolve(repoPath);

      if (!existsSync(gitDir)) {
        return false;
      }

      const repo = simpleGit(gitDir);
      // Try to access git directory to verify it's a valid repo
      await repo.revparse('--git-dir');
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Return metadata about the repository at the given path.
   *
   * @param repoPath - Root of a Git repository
   * @returns Object with branch, remoteUrl, and isDirty properties
   * @throws Error if the path is not a valid Git repository
   */
  async getRepoInfo(repoPath: string): Promise<RepoInfo> {
    const gitDir = path.resolve(repoPath);

    if (!existsSync(gitDir)) {
      throw new Error(`Repository path does not exist: ${gitDir}`);
    }

    const repo = simpleGit(gitDir);

    try {
      // Verify it's a valid git repo
      await repo.revparse('--git-dir');
    } catch {
      throw new Error(`Not a valid Git repository: ${gitDir}`);
    }

    // Get current branch
    let branch = '';
    try {
      const status = await repo.status();
      branch = status.current || 'detached';
    } catch {
      branch = 'unknown';
    }

    // Get remote URL
    let remoteUrl = '';
    try {
      const remotes = await repo.getRemotes(true);
      const originRemote = remotes.find((r) => r.name === 'origin');
      if (originRemote) {
        remoteUrl = originRemote.refs.fetch || '';
      }
    } catch {
      // Ignore error, remoteUrl remains empty
    }

    // Check if working directory is dirty
    let isDirty = false;
    try {
      const status = await repo.status();
      isDirty = !status.isClean();
    } catch {
      isDirty = false;
    }

    return {
      branch,
      remoteUrl,
      isDirty,
    };
  }

  /**
   * Return the count most recent commits from HEAD.
   *
   * @param repoPath - Root of a Git repository
   * @param count - Maximum number of commits to return (default 10)
   * @returns Array of commit objects with hash, message, author, and date
   * @throws Error if the path is not a valid Git repository
   */
  async getRecentCommits(repoPath: string, count = 10): Promise<Commit[]> {
    const gitDir = path.resolve(repoPath);

    if (!existsSync(gitDir)) {
      throw new Error(`Repository path does not exist: ${gitDir}`);
    }

    const repo = simpleGit(gitDir);

    try {
      // Verify it's a valid git repo
      await repo.revparse('--git-dir');
    } catch {
      throw new Error(`Not a valid Git repository: ${gitDir}`);
    }

    try {
      const log = await repo.log({ maxCount: count });
      return log.all.map((commit) => ({
        hash: commit.hash,
        message: commit.message.split('\n')[0], // First line only
        author: commit.author_name || 'unknown',
        date: new Date(commit.date).toISOString(),
      }));
    } catch (error) {
      // Empty repo or no commits
      if ((error as any)?.message?.includes('your current branch')) {
        return [];
      }
      throw error;
    }
  }
}
